File: leetcode/p00138/p00138.py
from typing import Optional

class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        
        memo = {}
        
        anchor = head
        root = Node(head.val)
        current = root

        memo[anchor] = root
        anchor = anchor.next

        while anchor:
            node = Node(anchor.val)
            current.next = node
            current = current.next

            memo[anchor] = current

            anchor = anchor.next
        
        anchor = head
        current = root

        while anchor:
            if anchor.random in memo:
                current.random = memo[anchor.random]
            
            anchor = anchor.next
            current = current.next
        
        return root

File: leetcode/p00138/p00138.go
package p00138

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}

	res := &Node{}

	origin, copyright := head, res
	mapping := map[*Node]*Node{}

	// 第一次遍歷 head 只處理兩件事情
	// 1. 複製出另一份與 head 一模一樣長度與 value 的 linked list
	// 2. 紀錄 linked list 長度中相同位置的新舊節點對照表
	for origin != nil {
		copyright.Val = origin.Val
		mapping[origin] = copyright

		// 末端位置邊界判斷
		if origin.Next != nil {
			copyright.Next = &Node{}
		}
		origin = origin.Next
		copyright = copyright.Next
	}

	// 第二次遍歷 head 按照對照表更新 linked list 的 random field
	origin, copyright = head, res
	for origin != nil {
		// 只處理確實有 random node 的節點
		if origin.Random != nil {
			copyright.Random = mapping[origin.Random]
		}
		origin = origin.Next
		copyright = copyright.Next
	}

	return res
}
